docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,306 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Super Simple Discovery API
+
+## Overview
+
+It is surprisingly hard to figure out how to talk to a Kubernetes cluster.  Not only do clients need to know where to look on the network, they also need to identify the set of root certificates to trust when talking to that endpoint.
+
+This presents a set of problems:
+* It should be super easy for users to configure client systems with a minimum of effort `kubectl` or `kubeadm init` (or other client systems).
+  * Establishing this should be doable even in the face of nodes and master components booting out of order.
+  * We should have mechanisms that don't require users to ever have to manually manage certificate files.
+* Over the life of the cluster this information could change and client systems should be able to adapt.
+
+## ClusterInfo object
+
+First we define a set of information that identifies a cluster and how to talk to it.
+
+This includes:
+
+* A cluster serial number/identifier.
+  * In an HA world, API servers may come and go and it is necessary to make sure we are talking to the same cluster as we thought we were talking to.
+  * It looks like Federation already has this concept.  We need to back port that to non-federated clusters.
+* A set of addresses for finding the cluster.
+  * It is implied that all of these are equivalent and that a client can try multiple until an appropriate target is found.
+  * Initially I’m proposing a flat set here.  In the future we can introduce more structure that hints to the user which addresses to try first.
+* A set of root certificates to trust for this cluster.
+  * We make this a set so that root certificates can be rotated.
+  * These are BASE64 encoded PEM values (without `-----BEGIN CERTIFICATE-----` lines).
+  This is similar to how [JWS encodes certifcates in JSON](https://www.google.com/url?q=https://tools.ietf.org/html/rfc7515%23appendix-B&sa=D&ust=1471374807953000&usg=AFQjCNFQNcmbquq3XEegKc76v5j956aSPw).
```

Comment:
Following through the links for this, it sounds like it's actually supposed to be a base64 encoded DER certificate. (https://tools.ietf.org/html/rfc7515#section-4.1.6)

This may not be a fun game to play if we're actually going to target that, a lot of hassle for no benefit on either end of the interaction. How do we feel about just returning the PEM including headers? (whole payload is base64 encoded, so once you decode and verify JWS you have a list of strings which are just certificate PEM) This would be a lot easier to work with.